File: scripts/council_verify.py
#!/usr/bin/env python3
"""Council run verifier — member-as-subagent enforcement (M1).

Validates a council run's artifact completeness under
``~/.ciel/council/<run_id>/`` and emits the run's structured member
verdicts to ``~/.ciel/improvements/signals/council-<run_id>.json`` so the
improvement loop can consume them.

Contract (see adapters/devin/COUNCIL_INVOCATION.md):
    spawn_receipts.json   {"mode": "subagent"|"inline", "members": {...}}
    members/<member>.stage1.json   {"member","stage":1,"score",...}
    members/<member>.stage2.json   {"member","stage":2,"score",...}
    verdict.json                   chairman docket with "votes" + "verdict"

``mode: "inline"`` is a declared degraded fallback — the run verifies but
is flagged ``unverified_member_isolation``. A run with no receipts at all
fails verification: member isolation is unproven.

Usage: council_verify.py <run_id|run_dir>
Exit: 0 verified (warnings allowed), 1 failed.
"""

import json
import sys
from pathlib import Path

MEMBERS = ["coherence", "capability", "safety", "efficiency", "evolution"]
CIEL = Path.home() / ".ciel"


def _load_json(path: Path) -> dict | None:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None


def verify(run_dir: Path) -> dict:  # noqa: PLR0912
    problems: list[str] = []
    warnings: list[str] = []

    receipts = _load_json(run_dir / "spawn_receipts.json")
    if receipts is None:
        problems.append("missing spawn_receipts.json — member isolation unproven")
        mode = "unknown"
    else:
        mode = receipts.get("mode", "unknown")
        if mode == "inline":
            warnings.append("unverified_member_isolation: run was inline, not subagent-dispatched")
        elif mode != "subagent":
            warnings.append(f"unrecognized spawn mode '{mode}'")

    member_verdicts = {}
    for member in MEMBERS:
        for stage in (1, 2):
            path = run_dir / "members" / f"{member}.stage{stage}.json"
            data = _load_json(path)
            if data is None:
                problems.append(f"missing/invalid {path.name}")
                continue
            if data.get("member") != member:
                problems.append(f"{path.name}: member field is '{data.get('member')}'")
            if data.get("stage") != stage:
                problems.append(f"{path.name}: stage field is {data.get('stage')}")
            score = data.get("score")
            if not isinstance(score, (int, float)):
                problems.append(f"{path.name}: non-numeric score {score!r}")
            else:
                member_verdicts.setdefault(member, {})[f"stage{stage}"] = score

    verdict = _load_json(run_dir / "verdict.json")
    if verdict is None:
        problems.append("missing/invalid verdict.json")
    else:
        votes = verdict.get("votes") or {}
        missing_votes = [m for m in MEMBERS if m not in votes]
        if missing_votes:
            problems.append(f"verdict.json missing votes for: {', '.join(missing_votes)}")
        safety = votes.get("safety")
        if verdict.get("verdict") == "pass" and isinstance(safety, (int, float)) and safety <= 3:
            problems.append("verdict=pass but safety <= 3 (veto condition)")

    return {
        "run_dir": str(run_dir),
        "mode": mode,
        "verified": not problems,
        "problems": problems,
        "warnings": warnings,
        "member_verdicts": member_verdicts,
    }


def emit_signal(run_id: str, result: dict) -> Path | None:
    signals = CIEL / "improvements" / "signals"
    try:
        signals.mkdir(parents=True, exist_ok=True)
        slug = run_id if run_id.startswith("council-") else f"council-{run_id}"
        out = signals / f"{slug}.json"
        out.write_text(json.dumps({
            "signal": "council_verdict",
            "run_id": run_id,
            "mode": result["mode"],
            "verified": result["verified"],
            "member_verdicts": result["member_verdicts"],
            "warnings": result["warnings"],
        }, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        return out
    except OSError:
        return None


def main() -> int:
    if len(sys.argv) != 2:
        print(__doc__.strip().splitlines()[0], file=sys.stderr)
        return 2
    arg = Path(sys.argv[1])
    run_dir = arg if arg.is_dir() else CIEL / "council" / sys.argv[1]
    if not run_dir.is_dir():
        print(f"[council_verify] no such run: {run_dir}")
        return 1

    result = verify(run_dir)
    run_id = run_dir.name
    signal_path = emit_signal(run_id, result)

    status = "VERIFIED" if result["verified"] else "FAILED"
    print(f"[council_verify] {run_id}: {status} (mode={result['mode']})")
    for w in result["warnings"]:
        print(f"  warning: {w}")
    for p in result["problems"]:
        print(f"  problem: {p}")
    if signal_path:
        print(f"  signal: {signal_path}")
    return 0 if result["verified"] else 1


if __name__ == "__main__":
    sys.exit(main())

File: ciel.skill/init/ciel-rs/src/councilverify.rs
//! `ciel council-verify <run_id|run_dir>` — Rust port of
//! `scripts/council_verify.py`. Validates a council run's artifact
//! completeness under `ciel_home()/council/<run_id>/` and emits the run's
//! structured member verdicts to `improvements/signals/council-<run_id>.json`.
//!
//! Contract (adapters/devin/COUNCIL_INVOCATION.md):
//!   spawn_receipts.json          {"mode": "subagent"|"inline", "members": {...}}
//!   members/<member>.stage{1,2}.json   {"member","stage":N,"score":num,...}
//!   verdict.json                 chairman docket with "votes" + "verdict"
//!
//! Exit: 0 verified (warnings allowed), 1 failed, 2 usage.

use serde_json::{json, Map, Value};
use std::path::{Path, PathBuf};

use crate::paths;

const MEMBERS: [&str; 5] = [
    "coherence",
    "capability",
    "safety",
    "efficiency",
    "evolution",
];

fn load_json(path: &Path) -> Option<Value> {
    serde_json::from_str(&std::fs::read_to_string(path).ok()?).ok()
}

pub fn verify(run_dir: &Path) -> Value {
    let mut problems: Vec<String> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();

    let receipts = load_json(&run_dir.join("spawn_receipts.json"));
    let mode = match &receipts {
        None => {
            problems.push("missing spawn_receipts.json — member isolation unproven".into());
            "unknown".to_string()
        }
        Some(r) => {
            let m = r
                .get("mode")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown")
                .to_string();
            if m == "inline" {
                warnings.push(
                    "unverified_member_isolation: run was inline, not subagent-dispatched".into(),
                );
            } else if m != "subagent" {
                warnings.push(format!("unrecognized spawn mode '{m}'"));
            }
            m
        }
    };

    let mut member_verdicts = Map::new();
    for member in MEMBERS {
        for stage in [1, 2] {
            let path = run_dir
                .join("members")
                .join(format!("{member}.stage{stage}.json"));
            let name = path
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default();
            let Some(data) = load_json(&path) else {
                problems.push(format!("missing/invalid {name}"));
                continue;
            };
            if data.get("member").and_then(|v| v.as_str()) != Some(member) {
                problems.push(format!(
                    "{name}: member field is '{}'",
                    data.get("member")
                        .map(|v| v.as_str().unwrap_or("null").to_string())
                        .unwrap_or_else(|| "null".into())
                ));
            }
            if data.get("stage").and_then(|v| v.as_i64()) != Some(stage as i64) {
                problems.push(format!(
                    "{name}: stage field is {}",
                    data.get("stage")
                        .map(|v| v.to_string())
                        .unwrap_or_else(|| "null".into())
                ));
            }
            match data.get("score") {
                // Python isinstance(x, (int,float)) — bool counts as int
                Some(s) if s.is_number() || s.is_boolean() => {
                    member_verdicts
                        .entry(member.to_string())
                        .or_insert_with(|| json!({}))
                        .as_object_mut()
                        .unwrap()
                        .insert(format!("stage{stage}"), s.clone());
                }
                other => problems.push(format!(
                    "{name}: non-numeric score {}",
                    other
                        .map(|v| v.to_string())
                        .unwrap_or_else(|| "null".into())
                )),
            }
        }
    }

    match load_json(&run_dir.join("verdict.json")) {
        None => problems.push("missing/invalid verdict.json".into()),
        Some(verdict) => {
            let votes = verdict.get("votes").cloned().unwrap_or_else(|| json!({}));
            let missing: Vec<&str> = MEMBERS
                .iter()
                .filter(|m| votes.get(**m).is_none())
                .copied()
                .collect();
            if !missing.is_empty() {
                problems.push(format!(
                    "verdict.json missing votes for: {}",
                    missing.join(", ")
                ));
            }
            let safety = votes.get("safety");
            let safety_num = safety.and_then(|s| {
                s.as_f64()
                    .or_else(|| s.as_bool().map(|b| if b { 1.0 } else { 0.0 }))
            });
            if verdict.get("verdict").and_then(|v| v.as_str()) == Some("pass")
                && safety_num.is_some_and(|n| n <= 3.0)
            {
                problems.push("verdict=pass but safety <= 3 (veto condition)".into());
            }
        }
    }

    json!({
        "run_dir": run_dir.to_string_lossy(),
        "mode": mode,
        "verified": problems.is_empty(),
        "problems": problems,
        "warnings": warnings,
        "member_verdicts": Value::Object(member_verdicts),
    })
}

fn emit_signal(ciel: &Path, run_id: &str, result: &Value) -> Option<PathBuf> {
    let signals = ciel.join("improvements").join("signals");
    std::fs::create_dir_all(&signals).ok()?;
    let slug = if run_id.starts_with("council-") {
        run_id.to_string()
    } else {
        format!("council-{run_id}")
    };
    let out = signals.join(format!("{slug}.json"));
    let body = json!({
        "signal": "council_verdict",
        "run_id": run_id,
        "mode": result["mode"],
        "verified": result["verified"],
        "member_verdicts": result["member_verdicts"],
        "warnings": result["warnings"],
    });
    let mut text = serde_json::to_string_pretty(&body).ok()?;
    text.push('\n');
    std::fs::write(&out, text).ok()?;
    Some(out)
}

/// `ciel council-verify <run_id|run_dir>`.
pub fn main_(args: &[String]) -> i32 {
    let ciel = paths::ciel_home();
    if args.len() != 1 {
        eprintln!("Council run verifier — member-as-subagent enforcement (M1).");
        return 2;
    }
    let arg = PathBuf::from(&args[0]);
    let run_dir = if arg.is_dir() {
        arg
    } else {
        ciel.join("council").join(&args[0])
    };
    if !run_dir.is_dir() {
        println!("[council_verify] no such run: {}", run_dir.display());
        return 1;
    }
    let result = verify(&run_dir);
    let run_id = run_dir
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();
    let signal_path = emit_signal(&ciel, &run_id, &result);

    let status = if result["verified"].as_bool().unwrap_or(false) {
        "VERIFIED"
    } else {
        "FAILED"
    };
    println!(
        "[council_verify] {run_id}: {status} (mode={})",
        result["mode"].as_str().unwrap_or("unknown")
    );
    if let Some(ws) = result["warnings"].as_array() {
        for w in ws.iter().filter_map(|w| w.as_str()) {
            println!("  warning: {w}");
        }
    }
    if let Some(ps) = result["problems"].as_array() {
        for p in ps.iter().filter_map(|p| p.as_str()) {
            println!("  problem: {p}");
        }
    }
    if let Some(sp) = signal_path {
        println!("  signal: {}", sp.display());
    }
    if result["verified"].as_bool().unwrap_or(false) {
        0
    } else {
        1
    }
}
